python/pyspark/sql/dataframe.py

```diff
@@ -1762,7 +1762,7 @@ def toPandas(self):
         else:
```

Comment:
If we wanted to check that a nullable int field actually has null values, we could do it here and then not change type it there are null values.  We would have to construct the pandas DataFrame first though.

```python
pdf = pd.DataFrame.from_records(self.collect(), columns=self.columns)

dtype = {}
for field in self.schema:
    if not(field.dataType == IntegerType and field.nullable and pdf[field.name].isnull().any()):
        pandas_type = _to_corrected_pandas_type(field.dataType)
        if pandas_type is not None:
              dtype[field.name] = pandas_type

for f, t in dtype.items():
         pdf[f] = pdf[f].astype(t, copy=False)
    return pdf
```
This does make a pass over the data to check though